pshtt/pshtt.py

```diff
@@ -217,75 +243,124 @@ def basic_check(endpoint):
+                return
+            except OpenSSL.SSL.Error as err:
+                endpoint.live = False
+                logging.warning("{}: Unexpected OpenSSL exception during retry.".format(endpoint.url))
+                utils.debug("{}: {}".format(endpoint.url, err))
+                return
+            except Exception as err:
+                endpoint.unknown_error = True
+                logging.warning("{}: Unexpected other unknown exception during requests retry.".format(endpoint.url))
+                utils.debug("{}: {}".format(endpoint.url, err))
+                return
+
+        # If it was a certificate error of any kind, it's live,
+        # unless SSLyze encounters a connection error later
         endpoint.live = True
 
-        # Figure out the error(s).
-        https_check(endpoint)
-
     except requests.exceptions.ConnectionError as err:
-        endpoint.live = False
-        utils.debug("{0}".format(err))
-        return
+        # We can get this for some endpoints that are actually live,
+        # so if it's https let's try sslyze to be sure
+        if(endpoint.protocol == "https"):
+            # https check later will set whether the endpoint is live
+            endpoint.https_full_connection = False
+        else:
+            endpoint.live = False
+        logging.warning("{}: Error connecting.".format(endpoint.url))
+        utils.debug("{}: {}".format(endpoint.url, err))
 
     # And this is the parent of ConnectionError and other things.
     # For example, "too many redirects".
     # See https://github.com/kennethreitz/requests/blob/master/requests/exceptions.py
     except requests.exceptions.RequestException as err:
         endpoint.live = False
-        logging.warn("Unexpected other requests exception.")
-        utils.debug("{0}".format(err))
+        logging.warning("{}: Unexpected other requests exception.".format(endpoint.url))
+        utils.debug("{}: {}".format(endpoint.url, err))
         return
 
     except Exception as err:
         endpoint.unknown_error = True
-        logging.warn("Unexpected other unknown exception during initial request.")
-        utils.debug("{0}".format(err))
+        logging.warning("{}: Unexpected other unknown exception during initial request.".format(endpoint.url))
+        utils.debug("{}: {}".format(endpoint.url, err))
         return
 
+    # Run SSLyze to see if there are any errors
+    if(endpoint.protocol == "https"):
+        https_check(endpoint)
+
+    if req is None:
+        return
+
+    # try to get IP address
+    # ip = req.raw._connection.sock.getpeername()
+    try:
+        if req.raw.closed is False:
+            ip = req.raw._connection.sock.socket.getpeername()[0]
+            if endpoint.ip is None:
+                endpoint.ip = ip
+            else:
+                if endpoint.ip != ip:
+                    utils.debug("{}: Endpoint IP is already {}, but requests IP is {}.".format(endpoint.url, endpoint.ip, ip))
+    except:
```

Comment:
@echudow Can you catch the actual exception or exceptions you are expecting here?  A bare `except` is kind of scary, and `flake8` won't accept it.

Also, should we log something (at least at the debug level) if we end up in the exception clause?  Even if not, there should be a comment explaining why it's not a problem if we get an exception and why the `pass` is the right thing to do here.

Reply:
@jsf9k I should be able to make it catch an actual exception (even if it's just of type Exception).  I'm just trying to do best effort to get the actual IP address that is being scanned.  However, it looks like requests doesn't record the IP information so I have to reach down to the socket, but in some cases the socket is already closed and it throws an exception.  I try to limit the exceptions by checking if it's already closed, but even still it doesn't want a bare except statement.  I should be able to fix it later tonight.

Reply:
Then just change it to catch `Exception` and add a comment with what you explained and I think it will be fine. :+1: